### PR TITLE
[CGSCC] Fix compile time blowup with large RefSCCs

### PR DESCRIPTION
**Pull request by aeubanks**
In some modules, e.g. Kotlin-generated IR, we end up with a huge RefSCC and the call graph updates done as a result of the inliner take a long time. This is due to RefSCC::removeInternalRefEdges() getting called many times, each time removing one function from the RefSCC, but each call to removeInternalRefEdges() is proportional to the size of the RefSCC.

There are two places that call removeInternalRefEdges(), in updateCGAndAnalysisManagerForPass() and LazyCallGraph::removeDeadFunction().

1) Since LazyCallGraph can deal with spurious (edges that exist in the graph but not in the IR) ref edges, we can simply not call removeInternalRefEdges() in updateCGAndAnalysisManagerForPass().

2) LazyCallGraph::removeDeadFunction() still ends up taking the brunt of compile time with the above change for the original reason. So instead we batch all the dead function removals so we can call removeInternalRefEdges() just once. This requires some changes to callers of removeDeadFunction() to not actually erase the function from the module, but defer it to when we batch delete dead functions at the end of the CGSCC run, leaving the function body as "unreachable" in the meantime. We still need to ensure that call edges are accurate. I had also tried deleting dead functions after visiting a RefSCC, but deleting them all at once at the end was simpler.

Many test changes are due to not performing unnecessary revisits of an SCC (the CGSCC infrastructure deems ref edge refinements as unimportant when it comes to revisiting SCCs, although that seems to not be consistently true given these changes) because we don't remove some ref edges. Specifically for devirt-invalidated.ll this seems to expose an inlining order issue with the inliner. Probably unimportant for this type of intentionally weird call graph.

Compile time: https://llvm-compile-time-tracker.com/compare.php?from=6f2c61071c274a1b5e212e6ad4114641ec7c7fc3&to=b08c90d05e290dd065755ea776ceaf1420680224&stat=instructions:u

### CODE REVIEW

**Review by nikic — COMMENTED** [review]
This looks reasonable to me, but I'm not really familiar with the LazyCallGraph infrastructure.

File: llvm/lib/Analysis/LazyCallGraph.cpp [hunk 1]
```diff
@@ -1515,57 +1511,68 @@ void LazyCallGraph::removeDeadFunction(Function &F) {
...
+    RefSCC *RC = lookupRefSCC(*N);
+    RCs[RC].push_back(N);
+  }
+  // Remove outgoing edges from all dead functions. Dead functions should
+  // already have had their call edges removed in markDeadFunction(), so we only
+  // need to worry about spurious ref edges.
+  for (auto [RC, DeadNs] : RCs) {
+    SmallVector<std::pair<Node *, Node *>> InternalEdgesToRemove;
+    for (Node *DeadN : DeadNs) {
+      if (lookupRefSCC(*DeadN) != RC)
```
Comment (alinas) [inline llvm/lib/Analysis/LazyCallGraph.cpp:1544]: Should this assert lookupRefSCC(*DeadN) == RC? 
I'm missing when this can happen.

**Review by alinas — APPROVED** [review]

File: llvm/lib/Analysis/LazyCallGraph.cpp (hunk 1 above)
Comment (aeubanks) [reply llvm/lib/Analysis/LazyCallGraph.cpp]: ah yes you're right, I've removed the check

**Comment by SvyatoslavScherbina** [comment]
Hi folks! Thank you, @aeubanks, for this improvement that fixed one of the blockers ([KT-68417](https://youtrack.jetbrains.com/issue/KT-68417)) for updating LLVM in Kotlin ([KT-49279](https://youtrack.jetbrains.com/issue/KT-49279)).
Thanks to @alinas as well, of course.

**Comment by wlei-llvm** [comment]
Hi @aeubanks , some of our services is broken with below assertion and we bisected to this commit, could you take a look? I think you probably needs a reduced repro from me, I can work on that, but just post here first in case you find some thing obvious to fix. Thanks in advance!

```
clang++: /home/wlei/local/upstream/llvm-project/llvm/lib/Analysis/CGSCCPassManager.cpp:1175: LazyCallGraph::SCC &updateCGAndAnalysisManagerForPass(llvm::LazyCallGraph &, LazyCallGraph::SCC &, LazyCallGraph::Node &, llvm::CGSCCAnalysisManager &, llvm::CGSCCUpdateResult &, llvm::FunctionAnalysisManager &, bool): Assertion `!UR.InvalidatedRefSCCs.count(RC) && "Invalidated the current RefSCC!"' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.	Program arguments: /home/wlei/local/upstream/llvm-build/bin/clang++ @/tmp/fbcc.89nih15w/compiler.argsfile
1.	<eof> parser at end of file
2.	Optimizer
 #0 0x00007f650c9fe948 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) /home/wlei/local/upstream/llvm-project/llvm/lib/Support/Unix/Signals.inc:723:13
 #1 0x00007f650c9fca20 llvm::sys::RunSignalHandlers() /home/wlei/local/upstream/llvm-project/llvm/lib/Support/Signals.cpp:106:18
 #2 0x00007f650c943dd8 (anonymous namespace)::CrashRecoveryContextImpl::HandleCrash(int, unsigned long) /home/wlei/local/upstream/llvm-project/llvm/lib/Support/CrashRecoveryContext.cpp:73:5
 #3 0x00007f650c943dd8 CrashRecoverySignalHandler(int) /home/wlei/local/upstream/llvm-project/llvm/lib/Support/CrashRecoveryContext.cpp:390:51
 #4 0x00007f650c03e6f0 __restore_rt (/lib64/libc.so.6+0x3e6f0)
 #5 0x00007f650c08b94c __pthread_kill_implementation (/lib64/libc.so.6+0x8b94c)
 #6 0x00007f650c03e646 gsignal (/lib64/libc.so.6+0x3e646)
 #7 0x00007f650c0287f3 abort (/lib64/libc.so.6+0x287f3)
 #8 0x00007f650c02871b _nl_load_domain.cold (/lib64/libc.so.6+0x2871b)
 #9 0x00007f650c037386 (/lib64/libc.so.6+0x37386)
#10 0x00007f650d50202e updateCGAndAnalysisManagerForPass(llvm::LazyCallGraph&, llvm::LazyCallGraph::SCC&, llvm::LazyCallGraph::Node&, llvm::AnalysisManager<llvm::LazyCallGraph::SCC, llvm::LazyCallGraph&>&, llvm::CGSCCUpdateResult&, llvm::AnalysisManager<llvm::Function>&, bool) /home/wlei/local/upstream/llvm-project/llvm/lib/Analysis/CGSCCPassManager.cpp:1176:3
#11 0x00007f650d4fdef4 llvm::updateCGAndAnalysisManagerForFunctionPass(llvm::LazyCallGraph&, llvm::LazyCallGraph::SCC&, llvm::LazyCallGraph::Node&, llvm::AnalysisManager<llvm::LazyCallGraph::SCC, llvm::LazyCallGraph&>&, llvm::CGSCCUpdateResult&, llvm::AnalysisManager<llvm::Function>&) /home/wlei/local/upstream/llvm-project/llvm/lib/Analysis/CGSCCPassManager.cpp:1190:10
#12 0x00007f650d4fdef4 llvm::CGSCCToFunctionPassAdaptor::run(llvm::LazyCallGraph::SCC&, llvm::AnalysisManager<llvm::LazyCallGraph::SCC, llvm::LazyCallGraph&>&, llvm::LazyCallGraph&, llvm::CGSCCUpdateResult&) /home/wlei/local/upstream/llvm-project/llvm/lib/Analysis/CGSCCPassManager.cpp:568:19
#13 0x00007f650c37bc4d llvm::detail::PassModel<llvm::LazyCallGraph::SCC, llvm::CGSCCToFunctionPassAdaptor, llvm::AnalysisManager<llvm::LazyCallGraph::SCC, llvm::LazyCallGraph&>, llvm::LazyCallGraph&, llvm::CGSCCUpdateResult&>::run(llvm::LazyCallGraph::SCC&, llvm::AnalysisManager<llvm::LazyCallGraph::SCC, llvm::LazyCallGraph&>&, llvm::LazyCallGraph&, llvm::CGSCCUpdateResult&) /home/wlei/local/upstream/llvm-project/llvm/include/llvm/IR/PassManagerInternal.h:90:5
#14 0x00007f650d4f970b llvm::PassManager<llvm::LazyCallGraph::SCC, llvm::AnalysisManager<llvm::LazyCallGraph::SCC, llvm::LazyCallGraph&>, llvm::LazyCallGraph&, llvm::CGSCCUpdateResult&>::run(llvm::LazyCallGraph::SCC&, llvm::AnalysisManager<llvm::LazyCallGraph::SCC, llvm::LazyCallGraph&>&, llvm::LazyCallGraph&, llvm::CGSCCUpdateResult&) /home/wlei/local/upstream/llvm-project/llvm/lib/Analysis/CGSCCPassManager.cpp:90:12
... 
```

**Comment by aeubanks** [comment]
@wlei-llvm yes a reduced repro would be helpful, I can't figure out what's wrong with just the assert

**Comment by wlei-llvm** [comment]
> @wlei-llvm yes a reduced repro would be helpful, I can't figure out what's wrong with just the assert

Hi @aeubanks , here is the reduced repro:

cmd:
```
opt -disable-output -passes="cgscc(devirt<4>(inline,openmp-opt-cgscc,function<>(simplifycfg<>)))" < reduced.ll
```

reduced.ll:
```
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128"
target triple = "x86_64-redhat-linux-gnu"
define internal fastcc void @foo() #0 personality ptr null {
  br label %1
1:                                                ; preds = %2, %0
  ret void
2:                                                ; No predecessors!
  call fastcc void @bar()
  br label %1
}
define fastcc void @baz() personality ptr null {
  switch i64 0, label %3 [
    i64 0, label %3
    i64 1, label %3
    i64 2, label %3
    i64 3, label %4
    i64 4, label %3
    i64 5, label %3
    i64 6, label %3
    i64 7, label %3
    i64 8, label %3
    i64 9, label %3
    i64 10, label %3
    i64 11, label %3
    i64 12, label %3
    i64 13, label %3
    i64 14, label %3
    i64 15, label %3
    i64 16, label %3
  ]
1:                                                ; No predecessors!
  %2 = landingpad { ptr, i32 }
          cleanup
  call fastcc void @foo()
  resume { ptr, i32 } zeroinitializer
3:                                                ; preds = %4, %0, %0, %0, %0, %0, %0, %0, %0, %0, %0, %0, %0, %0, %0, %0, %0, %0
  ret void
4:                                                ; preds = %0
  call fastcc void @baz()
  br label %3
}
define fastcc void @bar() personality ptr null {
  call fastcc void @baz()
  ret void
}
; uselistorder directives
uselistorder ptr null, { 3, 4, 0, 5, 6, 1, 7, 8, 2 }
attributes #0 = { "target-cpu"="haswell" }
!llvm.module.flags = !{!0}
!0 = !{i32 7, !"openmp", i32 50}
```
Please take a look, thanks!

**Comment by aeubanks** [comment]
I have a fix for the assert incoming

**Comment by wlei-llvm** [comment]
> #98213 should fix the assert, let me know if you are still seeing issues. sorry for the trouble (we really need to remove CallGraphUpdater...)

Thanks for the quick fix, I will test it on our services shortly.

**Comment by wlei-llvm** [comment]
> > #98213 should fix the assert, let me know if you are still seeing issues. sorry for the trouble (we really need to remove CallGraphUpdater...)
> 
> Thanks for the quick fix, I will test it on our services shortly.

I confirm it fixes the assertion errors on our services, thanks again!

**Pull request merged**